boards/nucleo_f446re/Makefile

```diff
@@ -0,0 +1,6 @@
+# Makefile for building the tock kernel for the NUCLEO-446RE platform
+#
+TARGET=thumbv7em-none-eabi
+PLATFORM=nucleo_f446re
+
+include ../Makefile.common
```

Comment:
How are you loading code onto the board?

Reply:
I am using `arm-none-eabi-objcopy --update-section` to create an ELF image with the apps.

```diff
diff --git a/boards/Makefile.common b/boards/Makefile.common
index c76f2422..24941be9 100644
--- a/boards/Makefile.common
+++ b/boards/Makefile.common
@@ -172,3 +172,4 @@ target/$(TARGET)/release/$(PLATFORM):
 target/$(TARGET)/debug/$(PLATFORM):
        $(Q)RUSTFLAGS="$(RUSTFLAGS_FOR_CARGO_LINKING)" $(CARGO) build $(VERBOSE) --target=$(TARGET)
        $(Q)$(SIZE) $@
+       arm-none-eabi-objcopy --update-section .apps=/home/tos-user/work/src/libtock-c/examples/buttons/build/cortex-m4/cortex-m4.tbf $@ $@-app
```
Then using gdb/openocd to load `nucleo_f446re-app` onto the board.

Reply:
Can you add those commands to the Makefile?

Reply:
I can add `flash` and `flash-debug` targets to `Makefile`. 

However, adding `program` target is going to be tricky for two reasons.

1. There is no tockloader support for this board.

2. Developing tockloader support is going to be challenging because the sector
   erase size is either 16KB, 64KB or 128KB. In hindsight, I feel I should have
   chosen a more modern STM32 board that supports 4KB or 2KB page erase.
         
In the `README.md`, I can document how to load programs onto the board, along
with a commented `program` target in `Makefile`.

Reply:
Just `flash` is fine, as long as there is some known way to load code.

And re: erase size, what does it depend on? 16K is pretty big, and we'd actually have to make some improvements to tockloader-proto, but it _should_ work (even if it won't be memory efficient...)

Reply:
> And re: erase size, what does it depend on? 16K is pretty big, and we'd
> actually have to make some improvements to tockloader-proto, but it _should_
> work (even if it won't be memory efficient...)

Erase size depends on MCU's flash memory controller. Some MCU's specify a page
erase, while others specify a sector erase.

In case of STM32F446RE, the sectors are organized as follows.

```
+----------+--------------------------+--------+
| Sector 0 | 0x0800_0000 - 0x0800_3FFF |  16KB |
+----------+---------------------------+-------+
| Sector 1 | 0x0800_4000 - 0x0800_7FFF |  16KB |
+----------+---------------------------+-------+
| Sector 2 | 0x0800_8000 - 0x0800_BFFF |  16KB |
+----------+---------------------------+-------+
| Sector 3 | 0x0800_C000 - 0x0800_FFFF |  16KB |
+----------+---------------------------+-------+
| Sector 4 | 0x0801_0000 - 0x0801_FFFF |  64KB |
+----------+---------------------------+-------+
| Sector 5 | 0x0802_0000 - 0x0803_FFFF | 128KB |
+----------+---------------------------+-------+
| Sector 6 | 0x0804_0000 - 0x0805_FFFF | 128KB |
+----------+---------------------------+-------+
| Sector 7 | 0x0806_0000 - 0x0807_FFFF | 128KB |
+----------+---------------------------+-------+
```

Hence we can only erase in blocks of either 16KB, 64KB or 128KB.

Reply:
Hm, that does make things a little complicated. I think it could be done (since writes don't have to be that large) but it would likely take a few tweaks.

Reply:
re `program`, please do add an empty target with explanation, similar to https://github.com/tock/tock/blob/master/boards/nordic/nrf51dk/Makefile#L27

We try to ensure that both `make program` and `make flash` "work", even if it's only meaningful to print directions on how to do the other thing in some cases